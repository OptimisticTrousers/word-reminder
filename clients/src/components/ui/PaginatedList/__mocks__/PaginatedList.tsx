import { Props } from "../PaginatedList";

export function PaginatedList(props: Props) {
  return <div data-testid="paginated-list">{String(props)}</div>;
}
