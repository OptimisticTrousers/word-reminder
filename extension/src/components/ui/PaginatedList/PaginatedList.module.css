.paginated__count {
  margin-bottom: 0.75rem;
}

.paginated__buttons {
  display: flex;
  width: 100%;
  gap: 0.5rem;
}

.paginated__button {
  flex: 1;
  font-size: 1.5rem;
}
