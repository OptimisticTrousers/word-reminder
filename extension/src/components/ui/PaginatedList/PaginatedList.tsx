import { ReactNode } from "react";
import CSSModules from "react-css-modules";
import { Link } from "react-router-dom";

import { ErrorMessage } from "../ErrorMessage";
import { Loading } from "../Loading";
import { NoMore } from "../NoMore";
import styles from "./PaginatedList.module.css";

interface Page {
  page: number;
  limit: number;
}

export interface Props {
  name: string;
  totalRows: string;
  list: ReactNode[];
  isLoading: boolean;
  isSuccess: boolean;
  error: Error | null;
  previous?: Page;
  next?: Page;
}

export const PaginatedList = CSSModules(
  function ({
    name,
    totalRows,
    list,
    isLoading,
    isSuccess,
    error,
    previous,
    next,
  }: Props) {
    return (
      <div styleName="paginated">
        <h2 styleName="paginated__heading">
          {name[0].toUpperCase() + name.slice(1)}
        </h2>
        <p styleName="paginated__count">{totalRows}</p>
        <ul styleName="paginated__list">
          {isSuccess && list.length === 0 ? <NoMore name={name} /> : list}
        </ul>
        {error && <ErrorMessage message={error.message} />}
        {isLoading && <Loading />}
        {previous && (
          <Link
            styleName="paginated__link"
            to={`/words?limit=${previous.limit}&page=${previous.page}`}
          >
            Previous
          </Link>
        )}
        {next && (
          <Link
            styleName="paginated__link"
            to={`/words?limit=${next.limit}&page=${next.page}`}
          >
            Next
          </Link>
        )}
      </div>
    );
  },
  styles,
  { allowMultiple: true, handleNotFoundStyleName: "log" }
);
