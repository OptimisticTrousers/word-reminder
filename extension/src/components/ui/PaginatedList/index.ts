export { PaginatedList } from "./PaginatedList";
export type { Props } from "./PaginatedList";
