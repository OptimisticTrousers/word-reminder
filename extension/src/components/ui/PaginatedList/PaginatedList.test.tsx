import { createRoutesStub } from "react-router-dom";
import { render, screen } from "@testing-library/react";

import { PaginatedList } from "./PaginatedList";

vi.mock("../ErrorMessage", function () {
  return {
    ErrorMessage: function ({ message }: { message: string }) {
      return (
        <p id="message" aria-labelledby="message">
          {message}
        </p>
      );
    },
  };
});

vi.mock("../Loading", function () {
  return {
    Loading: function () {
      return (
        <p id="loading" aria-labelledby="loading">
          Loading...
        </p>
      );
    },
  };
});

vi.mock("../NoMore", function () {
  return {
    NoMore: function ({ name }: { name: string }) {
      return (
        <p id="name" aria-labelledby="name">
          {name}
        </p>
      );
    },
  };
});

describe("PaginatedList component", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  const name = "words";
  const errorMessage = "Error message";
  const loadingMessage = "Loading...";

  describe("when it is successful", () => {
    const props = {
      name,
      totalRows: "3",
      list: [<li key={1}></li>, <li key={2}></li>, <li key={3}></li>],
      isLoading: false,
      isSuccess: true,
      error: null,
      previous: undefined,
      next: undefined,
    };

    it("renders list", async () => {
      const Stub = createRoutesStub([
        {
          path: `/${name}`,
          Component: function () {
            return <PaginatedList {...props} />;
          },
        },
      ]);

      const { asFragment } = render(<Stub initialEntries={[`/${name}`]} />);

      const heading = screen.getByRole("heading", { name: "Words" });
      const listItems = screen.getAllByRole("listitem", { name: "" });
      const emptyList = screen.queryByRole("paragraph", {
        name,
      });
      const error = screen.queryByRole("paragraph", { name: errorMessage });
      const loading = screen.queryByRole("paragraph", {
        name: loadingMessage,
      });
      const previousLink = screen.queryByRole("link", { name: "Previous" });
      const nextLink = screen.queryByRole("link", { name: "Next" });
      expect(heading).toBeInTheDocument();
      expect(listItems).toHaveLength(3);
      expect(loading).not.toBeInTheDocument();
      expect(emptyList).not.toBeInTheDocument();
      expect(error).not.toBeInTheDocument();
      expect(previousLink).not.toBeInTheDocument();
      expect(nextLink).not.toBeInTheDocument();
      expect(asFragment()).toMatchSnapshot();
    });

    it("renders that there are no things in the list when it is empty", () => {
      const props = {
        name: "words",
        totalRows: "0",
        list: [],
        isLoading: false,
        isSuccess: true,
        error: null,
        previous: undefined,
        next: undefined,
      };
      const Stub = createRoutesStub([
        {
          path: `/${name}`,
          Component: function () {
            return <PaginatedList {...props} />;
          },
        },
      ]);

      const { asFragment } = render(<Stub initialEntries={[`/${name}`]} />);

      const heading = screen.getByRole("heading", { name: "Words" });
      const listItems = screen.queryAllByRole("listitem", { name: "" });
      const emptyList = screen.getByRole("paragraph", {
        name,
      });
      const error = screen.queryByRole("paragraph", { name: errorMessage });
      const loading = screen.queryByRole("paragraph", { name: loadingMessage });
      const previousLink = screen.queryByRole("link", { name: "Previous" });
      const nextLink = screen.queryByRole("link", { name: "Next" });
      expect(heading).toBeInTheDocument();
      expect(listItems).toHaveLength(0);
      expect(loading).not.toBeInTheDocument();
      expect(emptyList).toBeInTheDocument();
      expect(error).not.toBeInTheDocument();
      expect(previousLink).not.toBeInTheDocument();
      expect(nextLink).not.toBeInTheDocument();
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("when it is loading", () => {
    it("renders that it is loading", async () => {
      const props = {
        name: "words",
        totalRows: "0",
        list: [],
        isLoading: true,
        isSuccess: false,
        error: null,
        previous: undefined,
        next: undefined,
      };
      const Stub = createRoutesStub([
        {
          path: `/${name}`,
          Component: function () {
            return <PaginatedList {...props} />;
          },
        },
      ]);

      const { asFragment } = render(<Stub initialEntries={[`/${name}`]} />);

      const heading = screen.getByRole("heading", { name: "Words" });
      const listItems = screen.queryAllByRole("listitem", { name: "" });
      const emptyList = screen.queryByRole("paragraph", {
        name,
      });
      const error = screen.queryByRole("paragraph", { name: errorMessage });
      const loading = screen.getByRole("paragraph", { name: loadingMessage });
      const previousLink = screen.queryByRole("link", { name: "Previous" });
      const nextLink = screen.queryByRole("link", { name: "Next" });
      expect(heading).toBeInTheDocument();
      expect(listItems).toHaveLength(0);
      expect(loading).toBeInTheDocument();
      expect(emptyList).not.toBeInTheDocument();
      expect(error).not.toBeInTheDocument();
      expect(previousLink).not.toBeInTheDocument();
      expect(nextLink).not.toBeInTheDocument();
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("when there is an error", () => {
    it("renders error message", async () => {
      const props = {
        name: "words",
        totalRows: "0",
        list: [],
        isLoading: false,
        isSuccess: false,
        error: new Error(errorMessage),
        previous: undefined,
        next: undefined,
      };
      const Stub = createRoutesStub([
        {
          path: `/${name}`,
          Component: function () {
            return <PaginatedList {...props} />;
          },
        },
      ]);

      const { asFragment } = render(<Stub initialEntries={[`/${name}`]} />);

      const heading = screen.getByRole("heading", { name: "Words" });
      const listItems = screen.queryAllByRole("listitem", { name: "" });
      const emptyList = screen.queryByRole("paragraph", {
        name,
      });
      const error = screen.getByRole("paragraph", { name: errorMessage });
      const loading = screen.queryByRole("paragraph", { name: loadingMessage });
      const previousLink = screen.queryByRole("link", { name: "Previous" });
      const nextLink = screen.queryByRole("link", { name: "Next" });
      expect(heading).toBeInTheDocument();
      expect(listItems).toHaveLength(0);
      expect(loading).not.toBeInTheDocument();
      expect(emptyList).not.toBeInTheDocument();
      expect(error).toBeInTheDocument();
      expect(previousLink).not.toBeInTheDocument();
      expect(nextLink).not.toBeInTheDocument();
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("pagination", () => {
    it("renders pagination links when there is a previous and a next page", async () => {
      const props = {
        name: "words",
        totalRows: "3",
        list: [<li key={1}></li>, <li key={2}></li>, <li key={3}></li>],
        isLoading: false,
        isSuccess: true,
        error: null,
        previous: { page: 1, limit: 3 },
        next: { page: 1, limit: 3 },
      };
      const Stub = createRoutesStub([
        {
          path: `/${name}`,
          Component: function () {
            return <PaginatedList {...props} />;
          },
        },
      ]);

      const { asFragment } = render(<Stub initialEntries={[`/${name}`]} />);

      const heading = screen.getByRole("heading", { name: "Words" });
      const listItems = screen.getAllByRole("listitem", { name: "" });
      const emptyList = screen.queryByRole("paragraph", {
        name,
      });
      const error = screen.queryByRole("paragraph", { name: errorMessage });
      const loading = screen.queryByRole("paragraph", { name: loadingMessage });
      const previousLink = screen.getByRole("link", { name: "Previous" });
      const nextLink = screen.getByRole("link", { name: "Next" });
      expect(heading).toBeInTheDocument();
      expect(listItems).toHaveLength(3);
      expect(loading).not.toBeInTheDocument();
      expect(emptyList).not.toBeInTheDocument();
      expect(error).not.toBeInTheDocument();
      expect(previousLink).toBeInTheDocument();
      expect(nextLink).toBeInTheDocument();
      expect(asFragment()).toMatchSnapshot();
    });

    it("renders the previous pagination link when there is a previous page but not a next page", async () => {
      const props = {
        name: "words",
        totalRows: "3",
        list: [<li key={1}></li>, <li key={2}></li>, <li key={3}></li>],
        isLoading: false,
        isSuccess: true,
        error: null,
        previous: { page: 1, limit: 3 },
        next: undefined,
      };
      const Stub = createRoutesStub([
        {
          path: `/${name}`,
          Component: function () {
            return <PaginatedList {...props} />;
          },
        },
      ]);

      const { asFragment } = render(<Stub initialEntries={[`/${name}`]} />);

      const heading = screen.getByRole("heading", { name: "Words" });
      const listItems = screen.getAllByRole("listitem", { name: "" });
      const emptyList = screen.queryByRole("paragraph", {
        name,
      });
      const error = screen.queryByRole("paragraph", { name: errorMessage });
      const loading = screen.queryByRole("paragraph", { name: loadingMessage });
      const previousLink = screen.getByRole("link", { name: "Previous" });
      const nextLink = screen.queryByRole("link", { name: "Next" });
      expect(heading).toBeInTheDocument();
      expect(listItems).toHaveLength(3);
      expect(loading).not.toBeInTheDocument();
      expect(emptyList).not.toBeInTheDocument();
      expect(error).not.toBeInTheDocument();
      expect(previousLink).toBeInTheDocument();
      expect(nextLink).not.toBeInTheDocument();
      expect(asFragment()).toMatchSnapshot();
    });

    it("renders the next pagination link when there is a next page but not a previous page", async () => {
      const props = {
        name: "words",
        totalRows: "3",
        list: [<li key={1}></li>, <li key={2}></li>, <li key={3}></li>],
        isLoading: false,
        isSuccess: true,
        error: null,
        previous: undefined,
        next: { page: 1, limit: 3 },
      };
      const Stub = createRoutesStub([
        {
          path: `/${name}`,
          Component: function () {
            return <PaginatedList {...props} />;
          },
        },
      ]);

      const { asFragment } = render(<Stub initialEntries={[`/${name}`]} />);

      const heading = screen.getByRole("heading", { name: "Words" });
      const listItems = screen.getAllByRole("listitem", { name: "" });
      const emptyList = screen.queryByRole("paragraph", {
        name,
      });
      const error = screen.queryByRole("paragraph", { name: errorMessage });
      const loading = screen.queryByRole("paragraph", { name: loadingMessage });
      const previousLink = screen.queryByRole("link", { name: "Previous" });
      const nextLink = screen.getByRole("link", { name: "Next" });
      expect(heading).toBeInTheDocument();
      expect(listItems).toHaveLength(3);
      expect(loading).not.toBeInTheDocument();
      expect(emptyList).not.toBeInTheDocument();
      expect(error).not.toBeInTheDocument();
      expect(previousLink).not.toBeInTheDocument();
      expect(nextLink).toBeInTheDocument();
      expect(asFragment()).toMatchSnapshot();
    });

    it("renders no pagination links when there is no previous or next page", async () => {
      const props = {
        name: "words",
        totalRows: "3",
        list: [<li key={1}></li>, <li key={2}></li>, <li key={3}></li>],
        isLoading: false,
        isSuccess: true,
        error: null,
        previous: undefined,
        next: undefined,
      };
      const Stub = createRoutesStub([
        {
          path: `/${name}`,
          Component: function () {
            return <PaginatedList {...props} />;
          },
        },
      ]);

      const { asFragment } = render(<Stub initialEntries={[`/${name}`]} />);

      const heading = screen.getByRole("heading", { name: "Words" });
      const listItems = screen.getAllByRole("listitem", { name: "" });
      const emptyList = screen.queryByRole("paragraph", {
        name,
      });
      const error = screen.queryByRole("paragraph", { name: errorMessage });
      const loading = screen.queryByRole("paragraph", { name: loadingMessage });
      const previousLink = screen.queryByRole("link", { name: "Previous" });
      const nextLink = screen.queryByRole("link", { name: "Next" });
      expect(heading).toBeInTheDocument();
      expect(listItems).toHaveLength(3);
      expect(loading).not.toBeInTheDocument();
      expect(emptyList).not.toBeInTheDocument();
      expect(error).not.toBeInTheDocument();
      expect(previousLink).not.toBeInTheDocument();
      expect(nextLink).not.toBeInTheDocument();
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
