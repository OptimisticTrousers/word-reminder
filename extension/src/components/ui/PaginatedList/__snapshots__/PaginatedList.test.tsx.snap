// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PaginatedList component > pagination > renders no pagination links when there is no previous or next page 1`] = `
<DocumentFragment>
  <div
    class="_paginated_78cfc4"
  >
    <h2
      class="_paginated__heading_78cfc4"
    >
      Words
    </h2>
    <p
      class="_paginated__count_78cfc4"
    >
      3
    </p>
    <ul
      class="_paginated__list_78cfc4"
    >
      <li />
      <li />
      <li />
    </ul>
  </div>
</DocumentFragment>
`;

exports[`PaginatedList component > pagination > renders pagination links when there is a previous and a next page 1`] = `
<DocumentFragment>
  <div
    class="_paginated_78cfc4"
  >
    <h2
      class="_paginated__heading_78cfc4"
    >
      Words
    </h2>
    <p
      class="_paginated__count_78cfc4"
    >
      3
    </p>
    <ul
      class="_paginated__list_78cfc4"
    >
      <li />
      <li />
      <li />
    </ul>
    <a
      class="_paginated__link_78cfc4"
      data-discover="true"
      href="/words?limit=3&page=1"
    >
      Previous
    </a>
    <a
      class="_paginated__link_78cfc4"
      data-discover="true"
      href="/words?limit=3&page=1"
    >
      Next
    </a>
  </div>
</DocumentFragment>
`;

exports[`PaginatedList component > pagination > renders the next pagination link when there is a next page but not a previous page 1`] = `
<DocumentFragment>
  <div
    class="_paginated_78cfc4"
  >
    <h2
      class="_paginated__heading_78cfc4"
    >
      Words
    </h2>
    <p
      class="_paginated__count_78cfc4"
    >
      3
    </p>
    <ul
      class="_paginated__list_78cfc4"
    >
      <li />
      <li />
      <li />
    </ul>
    <a
      class="_paginated__link_78cfc4"
      data-discover="true"
      href="/words?limit=3&page=1"
    >
      Next
    </a>
  </div>
</DocumentFragment>
`;

exports[`PaginatedList component > pagination > renders the previous pagination link when there is a previous page but not a next page 1`] = `
<DocumentFragment>
  <div
    class="_paginated_78cfc4"
  >
    <h2
      class="_paginated__heading_78cfc4"
    >
      Words
    </h2>
    <p
      class="_paginated__count_78cfc4"
    >
      3
    </p>
    <ul
      class="_paginated__list_78cfc4"
    >
      <li />
      <li />
      <li />
    </ul>
    <a
      class="_paginated__link_78cfc4"
      data-discover="true"
      href="/words?limit=3&page=1"
    >
      Previous
    </a>
  </div>
</DocumentFragment>
`;

exports[`PaginatedList component > when it is loading > renders that it is loading 1`] = `
<DocumentFragment>
  <div
    class="_paginated_78cfc4"
  >
    <h2
      class="_paginated__heading_78cfc4"
    >
      Words
    </h2>
    <p
      class="_paginated__count_78cfc4"
    >
      0
    </p>
    <ul
      class="_paginated__list_78cfc4"
    />
    <div
      data-testid="loading"
    >
      Loading...
    </div>
  </div>
</DocumentFragment>
`;

exports[`PaginatedList component > when it is successful > renders list 1`] = `
<DocumentFragment>
  <div
    class="_paginated_78cfc4"
  >
    <h2
      class="_paginated__heading_78cfc4"
    >
      Words
    </h2>
    <p
      class="_paginated__count_78cfc4"
    >
      3
    </p>
    <ul
      class="_paginated__list_78cfc4"
    >
      <li />
      <li />
      <li />
    </ul>
  </div>
</DocumentFragment>
`;

exports[`PaginatedList component > when it is successful > renders that there are no things in the list when it is empty 1`] = `
<DocumentFragment>
  <div
    class="_paginated_78cfc4"
  >
    <h2
      class="_paginated__heading_78cfc4"
    >
      Words
    </h2>
    <p
      class="_paginated__count_78cfc4"
    >
      0
    </p>
    <ul
      class="_paginated__list_78cfc4"
    >
      <p
        aria-labelledby="name"
        id="name"
      >
        words
      </p>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`PaginatedList component > when there is an error > renders error message 1`] = `
<DocumentFragment>
  <div
    class="_paginated_78cfc4"
  >
    <h2
      class="_paginated__heading_78cfc4"
    >
      Words
    </h2>
    <p
      class="_paginated__count_78cfc4"
    >
      0
    </p>
    <ul
      class="_paginated__list_78cfc4"
    />
    <div
      data-testid="error-message"
    >
      Error message
    </div>
  </div>
</DocumentFragment>
`;
